RFC Function Contracts (#2620)
model-checking/kani: rfc/src/rfcs/0008-function-contracts.md

- **Feature Name:** Function Contracts
- **Feature Request Issue:** -
- **RFC PR:** [#2620](https://github.com/model-checking/kani/pull/2620)
- **Status:** Under Review 
- **Version:** 0
- **Proof-of-concept:** [features/contracts](https://github.com/model-checking/kani/tree/features/contracts)

-------------------

## Summary

> Function contracts are a mechanism for a stubbing-like abstraction of concrete

Reviewer: I'm trying to understand why you are saying "paves the way" for scalability and "lays the ground for modular verification". This RFC is intended to enable modular verification and improve scalability.

implementations but with a significantly reduced threat to soundness[^simple-unsoundness].

Lays the ground work for modular verification.
<!-- Shorter? -->

## User Impact

> <!-- Is basically the pitch and addressing the user. -->

Reviewer: more like a note to your note. :)

Function contracts provide an interface for verified abstraction, a mechanism
similar to stubbing, but sound[^simple-unsoundness]. It also paves the way for
the following two ambitious goals.

[^simple-unsoundness]: The main remaining threat to soundness in the use of
    contracts, as defined in this proposal, is the reliance on user-supplied
    harnesses for contract checking (explained in bulled 2 of [user
    experience](#user-experience)). A more thorough discussion on the dangers
    and potential remedies can be found in the [future](#future-possibilities)
    section.

- **Scalability:** Function contracts are sound (over)abstractions of function
  behavior. By verifiying the contract against its implemetation and
  subsequently performing caller verification against the (cheaper) abstraction
  verification can be modularized, cached and thus scaled.
- **Unbounded Verification:** Contracts can be reasoned over inductively and
  thus verify recursive functions.

> Enabling function contracts is a non-invasive change. While it contains a new

Reviewer: I mean this sentence is basically trying to say that this is backward compatible or that contracts are completely optional, right? And the usage of contract as stubbing has to be explicitly declared.

> Enabling function contracts is a non-invasive change. While it contains a new

Author: Rephrased

> Enabling function contracts is a non-invasive change. While it contains a new

Reviewer: Added a new comment to your new version.

   Specifically the domain (inputs) of the function described (by the `requires`
   clause) must be *at most* as large as the input space for which the
   function's behavior is well defined, and the codomain (outputs, described by
   `ensures`) must be *at least* as large as the actual space of outputs the
   function may produce.

   For example in our case it would be permissible to use
   `#[kani::requires(divisor > 100)]` (smaller permissible input domain) or
   `#[kani::ensures(result < dividend + divisor)]` (larger possible output
   domain), but `#[kani::ensures(result < dividend)]` (small output) is not
   allowed.

   To facilitate the check Kani needs a suitable environment to verify the
   function in. For this proposal the environment must be provided by us (the
   users). See [future possibilities](#future-possibilities) for a discussion
   about the arising soundness issues and their remedies.

   We provide the checking environment for our contract with a special new
   `proof_for_contract` harness.

   ```rs
   #[kani::proof_for_contract(my_div)]
   fn my_div_harness() {
     my_div(kani::any(), kani::any())
   }
   ```

>    Similar to a unit-test harness for any other function, we are supposed to

Reviewer: Maybe just mention that users must ensure that these harnesses provide all possible input combinations to the function. I think this is more relevant when handling references.

Also, maybe say "verification harness" instead of 'unit-test harness". I read the latter as concrete tests, not symbolic.

   make sure we catch all edge cases, then call the function at least once with
   those abstract inputs. If we forget to call `my_div` Kani would report an
   error.
   
   Unlike a unit-test we can however elide any checks of the output and
   post-call state. Instead Kani uses the conditions we specified in the
   contract as checks. Preconditions (`requires`) are inserted as `kani::assume`
   *before* the call to `my_div`, limiting the generic inputs to those the
   function is actually defined for. Postconditions (`ensures`) are inserted as
   `kani::assert` checks *after* the call to `my_div`.

   The expanded version of our harness and function is equivalent to the following:

   ```rs
   #[kani::proof]
   fn my_div_harness() {
     let dividend = kani::any();
     let divisor = kani::any();
     kani::assume(divisor != 0);
     let result = my_div(dividend, divisor);
     kani::assert(result <= dividend);
   }
   ```

   This expanded harness is then verified like any other harness but also gives
   the green light for the next step: verified stubbing.

3. In the last phase the **verified** contract is ready for us to use to
   **stub** other harnesses.

>    Unlike in regular stubbing Kani there to be at least one associated

Reviewer: ```suggestion
   Unlike in regular stubbing, there has to be at least one associated
```

Reviewer: Also, please add a link to the stub RFC.

>    `proof_for_contract` harness for each function to stub *and* it requires all

Reviewer: Are we actually planning on enforcing this? What about dependencies?

Author: I worry that contracts are easily used unsoundly (e.g. unchecked). Maybe an opt-out mechanism is the way to go. I think in the end this should be fixed with a cache that is discussed later

   such harnesses to pass verification before attempting verification of any
   harnesses that use it as a stub.

   A possible harness that uses our `my_div` contract could be the following:

   ```rs
   #[kani::proof]
   #[kani::stub_verified(my_div)]
   fn use_div() {
     let v = vec![...];
     let some_idx = my_div(v.len() - 1, 3);
     v[some_idx];
   }
   ```

   To use the contract as a stub Kani must first ensure the calling context is
   safe. It inserts a `kani::assert` for the preconditions (`requires`) before
   the call; then it replaces the result of `my_div` with a non-deterministic
   value. Finally the non-deterministic result is constrained by a
   `kani::assume` of the postconditions (`ensures`).
   
   Mutable memory is similarly made non-deterministic, discussed later in
   [havociing](#memory-predicates-and-havocing).

>    The expanded version is equivalent to this:

Reviewer: What if you show as the expanded version just the implementation of my_div?

```rust
fn my_div_stub(dividend: u32, divisor: u32) -> u32 {
     kani::assert(divisor != 0); // pre-condition
     kani::any_where(|result| { /* post-condition */ result <= dividend })
}
```

  
   ```rs
   #[kani::proof]
   #[kani::stub_verified(my_div)]
   fn use_div() {
     let v = vec![...];
     let dividend = v.len() - 1;
     let divisor = 3;
     kani::assert(divisor != 0);
     let result = kani::any();
     kani::assume(result <= dividend);
     let some_idx = result;
     v[some_idx];
   }
   ```

   Notice that this performs no actual computation for `my_div` (other than the
   conditions) which allows us to avoid something potentially costly.

Also notice that Kani was able to express both contract checking and stubbing
with existing capabilities; the important feature is the enforcement. The
checking is, by construction, performed **against the same condition** that is
later used as stub, which ensures soundness (see discussion on lingering threats
to soundness in the [future](#future-possibilities) section) and guarding against
stubs diverging from their checks.

### History Variables

> Kani's contract language contains additional support for reasoning about changes

Reviewer: I would remove this first line. History variables are also relevant for parameters that are passed by value even if the function doesn't have a side effect.

I think this paragraph is more relevant for explaining assign clauses.

to memory. One case where this is necessary is whenever `ensures` needs to
reason about state before the function call. By default it only has access to
state after the call completes, which will be different if the call mutates
memory.

Consider the `Vec::pop` function

```rs
impl<T> Vec<T> {
  fn pop(&mut self) -> Option<T> {
    ...
  }
}
```

If we want to describe in which case the result is `Some`, we need to know
whether `self` is empty *before* `pop` is called. To do this Kani provides the
`old(EXPR)` pseudo function, which evaluates `EXPR` before the call (e.g. to
`pop`) and makes the result available to `ensures`. It is used like so:

```rs
impl<T> Vec<T> {

>   #[kani::ensures(old(self.is_empty()) || result.is_some())]

Reviewer: Question, can one write this as:
```rust
#[kani::ensures(if old(self.is_empty()) { result.is_none() })]
```

  fn pop(&mut self) -> Option<T> {
    ...
  }
}
```

`old` allows evaluating any (side-effect free[^side-effects]) Rust expression.

> The bowwor checker enforces the result of `old` cannot observe the mutations

Reviewer: ```suggestion
The borrow checker enforces the result of `old` cannot observe the mutations
```

from e.g. `pop`, as that would defeat the purpose. If `your` expression in `old`

> returns borrowed content, make a copy instead (using e.g. `clone()`).

Reviewer: I'm confused by this explanation... I'm guessing that you are trying to say is that `old(self).is_empty()` is not equivalent to `old(self.is_empty())`.

Since `self` is a reference, `old(self)` will only refer to the value of the reference, not the value of the object pointed to.

Note also that `old` is syntax, not a function and implemented as an extraction
and lifting during code generation. It can reference e.g. `pop`'s arguments but
not local variables. Compare the following

**Invalid ❌:** `#[kani::ensures({ let x = self.is_empty(); old(x) } || result.is_some())]`</br>
**Valid ✅:** `#[kani::ensures(old({ let x = self.is_empty(); x }) || result.is_some())]`

And it will only be recognized as `old(...)`, not as `let old1 = old; old1(...)` etc.

### Memory Predicates and Havocing

> The last new feature added are predicates to refine a function's access to heap

Reviewer: I would prefer if we restructure this section.
1. Talk about function side effects at the top. Mention that the last part of a function specification is to determine what kind of side effects this function can trigger. Mention mutable static variables, mutable references and interior mutability.
2. Mention that we will add `#[kani::assign]` to allow users to specify which side effects a function produces.
3. Mention that Kani will assume that all mutable references and interior mutable elements can be modified, unless there's an existing assigns clause that narrows down the scope of the modifications.
4. Mention that changes to mutable static variables must be explicitly specified.

memory. A memory footprint is used by the verifier to perform "havocing" during
contract stubbing. Recall that stubbing replaces the result value with a
non-deterministic `kani::any()`, havocing is the equivalent memory regions
touched by the function. Any memory regions in the footprint are "havoced" by
the verifier, that is replaced by a non-deterministic value (subject to type
constraints).

By default Kani infers a memory footprint as all memory reachable from a `&mut`
input or any `static` global referenced, directly or transitively, by the
function. While the inferred footprint is sound and enough for successful
contract checking[^inferred-footprint] it can easily turn large section of
memory to non-deterministic values, invalidate invariants of your program and
cause the verification to fail when the contract is used as a stub.

[^inferred-footprint]: While inferred memory footprints are sound for both safe
    and unsafe rust certain features in unsafe rust (e.g. `RefCell`) get
    inferred incorrectly and will lead to a failing contract check.

> To reduce the scope of havocing Kani provides the `#[kani::assigns(CONDITION,

Reviewer 2: This section needs an example to walk the reader through it as well.

ranges of pointer offsets. A slicing syntax `p[i..j]` only applies if `p` is a
`*mut T` and points to an array allocation. The slice indices are offsets with
sizing `T`, e.g. in Rust `p[i..j]` would be equivalent to
`std::slice::from_raw_parts(p.offset(i), i - j)`. `i` must be smaller or equal
than `j`.

Because lvalues are restricted to using projections only, Kani must break
encapsulation here. If need be we can reference fields that are usually hidden,
without an error from the compiler.

[^side-effects]: Code used in contracts is required to be side effect free which
    means it must not perform I/O, mutate memory (`&mut` vars and such) or
    (de)allocate heap memory. This is enforced by the verifier, see the
    discussion in the [future](#future-possibilities) section.

### Detailed Attribute Contraints Overview

Any violation of the following constraints constitutes a compile-time error.

- A function may have any number of `requires`, `ensures`, `assigns` and `frees`
  attributes. Any function with at least one such annotation is considered as
  "having a contract".

  Harnesses (general or for contract checking) may not have any such annotation.

> - A function may have one `proof_for_contract(TARGET)` annotation. `TARGET` must

Reviewer: ```suggestion
- A harness may have up to one `proof_for_contract(TARGET)` annotation where `TARGET` must
```

>   "have a contract". Multiple functions may have `proof_for_contract`

Reviewer: One or more harnesses can target the same function.

>   annotations with the same `TARGET`. All such harnesses must pass verification,

Reviewer: I am not convinced we should enforce this. Maybe we can enforce this for contracts defined in the same crate as the stub_verified is used, but this will be much harder for dependency crates.

I also think this will make `--harness` behavior very awkward, since users won't be able to run their harnesses in isolation.

  before `TARGET` may be used as a verified stub.

>   A `proof_for_contract` harness may use both `stub` and `stub_verified`, though

Reviewer: I would just say that a `proof_for_contract` harness can have all attributes that are applicable to a regular harness.

>   A `proof_for_contract` harness may use both `stub` and `stub_verified`, though

Reviewer: ```suggestion
-  A `proof_for_contract` harness may use both `stub` and `stub_verified`, though
```

  the `TARGET` may not appear in either. 

>   Kani checks that `TARGET` is reachable from the `proof_for_contract` harness,

Reviewer: ```suggestion
-  Kani checks that `TARGET` is reachable from the `proof_for_contract` harness,
```

  but it does not warn if stubbed functions use `TARGET`[^stubcheck].

>   A `proof_for_contract` function may not have the `kani::proof` attribute (it

Reviewer: ```suggestion
-  A `proof_for_contract` function may not have the `kani::proof` attribute (it
```

  `kani::result()`, though with the latter I worry that people may get confused
  because it is syntactic and not subject to usual `use` renaming and import
  semantics. Alternatively we can let the user pick the name with an additional
  argument to `ensures`, e.g. `ensures(my_result_var, CONDITION)`

 
<!-- 
- Is there any part of the design that you expect to resolve through the RFC process?
- What kind of user feedback do you expect to gather before stabilization? How will this impact your design? 
-->

## Future possibilities

<!-- For Developers -->

- **Quantifiers:** Quantifiers are like logic-level loops and a powerful
  reasoning helper. CBMC has support for both `exists` and `forall`, but the
  code generation is difficult. The most ergonomic and easy way to implement
  quantifiers on the Rust side is as higher-order functions taking `Fn(T) ->
  bool`, where `T` is some arbitrary type that can be quantified over. This
  interface is familiar to developers, but the code generation is tricky, as
  CBMC level quantifiers only allow certain kinds of expressions. This
  necessiates a rewrite of the `Fn` closure to a compliant expression.

> - **Side effect** freedom is currently enforced by CBMC. This means that the error

Author: Added to scope

    harnesses for contract checking (explained in bulled 2 of [user
    experience](#user-experience)). A more thorough discussion on the dangers
    and potential remedies can be found in the [future](#future-possibilities)
    section.

- **Scalability:** Function contracts are sound (over)abstractions of function
  behavior. By verifiying the contract against its implemetation and
  subsequently performing caller verification against the (cheaper) abstraction,
  verification can be modularized, cached and thus scaled.
- **Unbounded Verification:** The abstraction provided by the contract can be used instead of a recursive call, thus allowing verification of recursive functions.

Enabling function contracts is a non-invasive change. While it contains a new
API, that API is strictly additive for users. All prior operations are unchanged.

### Caveats

We add function contract functionality, but do not add the inductive reasoning
support needed for many unbounded problems, such as ["decreases"
measures](https://diffblue.github.io/cbmc/contracts-decreases.html) and
inductive lemmas (for instance in [ACSL](https://frama-c.com/download/acsl.pdf)

> section 2.6.3 "inductive predicates").

Author: Recursion section added. I tried to be more specific on this, but truth be told I haven't had the chance to experiment with what we can achieve without inductive lemmas.

   }
   ```
  
   `requires` here indicates this function expects its `divisor` input to never
   be 0, or it will not execute correctly (i.e. panic).

   `ensures` puts a bound on the output, relative to the `dividend` input.

   Conditions in contracts are plain Rust expressions which can reference the
   function arguments and, in case of `ensures`, the result of the function as a
   special `result`[^result-naming] variable. Syntactically Kani supports any
   Rust expression, including function calls, defining types etc. However they
   must be side-effect free[^side-effects] or Kani will throw a compile error.

   Multiple `requires` and `ensures` clauses are allowed on the same function,
   they are implicitly logically conjoined.

   [^result-naming]: See [open questions](#open-questions) for a discussion
       about naming of the result variable.

2. Next Kani must make sure that the approximation we specified actually holds
   by **checking** the contract against the implementation. This is in contrast
   to ["stubbing"][stubbing], where the approximation is blindly trusted.

   The contract must always overapproximate the function to guarantee soundness.

>    To facilitate the check Kani needs a suitable environment to verify the

Author: They are the same. The general check if the function contract is checking that it overapproximates.

   function in. For this proposal the environment must be provided by us (the
   users). See [future possibilities](#future-possibilities) for a discussion
   about the arising soundness issues and their remedies.

   We provide the checking environment for our contract with a special new
   `proof_for_contract` harness.

   ```rs
   #[kani::proof_for_contract(my_div)]
   fn my_div_harness() {
     my_div(kani::any(), kani::any())
   }

>    ```

Reviewer: In theory yes, we should be able to check contracts for any harness that contains a contracts clause. In practice, we can't yet. We could make automatic harness generation a requirement for stabilization (at least for now) and the harness for contracts a temporary work around? What do you think @adpaco-aws?

>    ```

Author: Because heaps are complicated. This is what Remi was working on and hasn't yet finished. Basically it's hard to create arbitrary heap datastructures that you know will exercise all behaviors of the function you want to check.

>    ```

Reviewer 3: That's a technical limitation which doesn't seem to be mentioned anywhere. However, it looks like we're working around it with sacrifices to UX. Wouldn't it be better to reduce the scope and go for the optimal UX?

In other words, if we can detect the cases where we need arbitrary heap data structures, we can refuse contract checking/replacement. That way, UX doesn't need to change when the work there is completed, we can just remove the mitigation (detection).

>    ```

Reviewer 3: Resolved offline.

>    ```

Author: I've added this discussion to the "Rational" section.

Imo the limitations this UX would impose is unreasonable. I'd rather enable users to be productive and gradually improve their UX.

   *before* the call to `my_div`, limiting the generic inputs to those the
   function is actually defined for. Postconditions (`ensures`) are inserted as
   `kani::assert` checks *after* the call to `my_div`.

   The expanded version of our harness and function is equivalent to the following:

   ```rs
   #[kani::proof]
   fn my_div_harness() {
     let dividend = kani::any();
     let divisor = kani::any();
     kani::assume(divisor != 0); // requires
     let result = my_div(dividend, divisor);
     kani::assert(result <= dividend); // ensures
   }
   ```

   This expanded harness is then verified like any other harness but also gives
   the green light for the next step: verified stubbing.

3. In the last phase the **verified** contract is ready for us to use to
   **stub** other harnesses.

   Unlike in regular stubbing, there has to be at least one associated
   `proof_for_contract` harness for each function to stub *and* it requires all
   such harnesses to pass verification before attempting verification of any
   harnesses that use it as a stub.

   A possible harness that uses our `my_div` contract could be the following:

   ```rs
   #[kani::proof]

>    #[kani::stub_verified(my_div)]

Reviewer 3: Not sure we want to call this `stub_*`, we should try to avoid users confusing stubbing and function contracts.

Reviewer: Interesting, that was my suggestion. This annotation is telling Kani to use the contract as a stub, so what would it be your suggestion? I think I initially suggested something like `stub_with_contract` but it is rather long.

Reviewer 3: I'd suggest `use_contract`

Author: Well but what does "using" a contract entail? It's a bit ambiguous. You're could argue that you're "using" it when checking, no?

Reviewer 3: Yes, it's a little ambiguous since "use" is auxiliary. Won't push on the specific name anymore.

   the call; then it replaces the result of `my_div` with a non-deterministic
   value. Finally the non-deterministic result is constrained by a
   `kani::assume` of the postconditions (`ensures`).
   
   Mutable memory is similarly made non-deterministic, discussed later in
   [havociing](#memory-predicates-and-havocing).

   Stubbing `my_div` expands it similar to this:
  
   ```rs
   fn my_div_stub(dividend: u32, divisor: u32) -> u32 {
     kani::assert(divisor != 0); // pre-condition
     kani::any_where(|result| { /* post-condition */ result <= dividend })
   }
   ```

   Notice that this performs no actual computation for `my_div` (other than the
   conditions) which allows us to avoid something potentially costly.

Also notice that Kani was able to express both contract checking and stubbing
with existing capabilities; the important feature is the enforcement. The
checking is, by construction, performed **against the same condition** that is
later used as stub, which ensures soundness (see discussion on lingering threats
to soundness in the [future](#future-possibilities) section) and guarding against
stubs diverging from their checks.

> 

Reviewer 3: This section describes a lot of low-level details (e.g., the expansions) which in my opinion are more appropriate to discuss in the design section. However, what I'm missing here is the justification to have new APIs and even contract-specific harnesses when there's apparently no need to. Or did I miss something?

Author: I think mentioning the expansion was @feliperodri's wish if I recall correctly and I think you have a differing conception of what this section is supposed to contain. I was told this is really addressing the user who cares about how to use the contracts and maybe understanding a bit of how they work. They wouldn't care so much as to the justification for extra harnesses.

I'm not sure what you mean by "new APIs". What specifically are you referring to?

The contract specific harnesses exist because we need *some* way of saying which harness tests which contract and we chose this one. There is a section in "Rationale and Alternatives" discussing the justification which I have yet to fill in. Sorry

Reviewer 3: > I think you have a differing conception of what this section is supposed to contain.

Fair enough.

> What specifically are you referring to?

I'm guessing I was referring to the `proof_for_contract` and `stub_verified` attributes. But let me continue this discussion on the other thread.

[^side-effects]: Code used in contracts is required to be side effect free which
    means it must not perform I/O, mutate memory (`&mut` vars and such) or
    (de)allocate heap memory. This is enforced by the verifier, see the
    discussion in the [future](#future-possibilities) section.

> ### Write Sets and Havocing

Reviewer 3: I've stopped reviewing here. Note that "havocing" -> "havocking".

too coarse grained. In this case every value in this vector will be made
non-deterministic.

This proposal also adds an `assigns` and `frees` clause which limits the scope
of havocing. Both clauses represent an assertion that the function will modify
only the specified memory regions. Similar to requires/ensures the verifier
enforces the assertion in the checking stage to ensure soundness. When the
contract is used as a stub the assigns clause is used as the write set to havoc.

In our `pop` example the only modified memory location is the last element and
only if the vector was not already empty, which would be specified thusly.

```rs
impl<T> Vec<T> {
  #[assigns(!self.is_empty(), (*self).buf.ptr.pointer.pointer[self.len])]
  #[assigns(self.is_empty(), false)]
  fn pop(&mut self) -> Option<T> {
    ...
  }
}
```

The `#[assigns(CONDITION, ASSIGNS_RANGE)]` consists of an optional `CONDITION`
(defaults to `true`) and a `ASSIGNS_RANGE` which is essentially an lvalue. 

Lvalues are simple expressions permissible on the left hand side of an
assignment. They compose of the name of one function argument and zero or more

> projections (dereference `*`, field access `.x`, slice indexing `[1]`).

Author: I'm limiting it to lvalues, constants and arithmetic for now

approximations for modular verification which affords both scalability, but also
allows for verifying unbounded loops and recursion.

<!-- Shorter? -->

## User Impact

<!-- Is basically the pitch and addressing the user. -->

Function contracts are a mechanism for a stubbing-like abstraction of concrete
implementations but with a significantly reduced threat to soundness. It also
lays the ground work for the following two ambitious goals.

- **Scalability:** Function contracts are sound (over)abstractions of function
  behavior. By verifiying the contract against its implemetation and
  subsequently performing caller verification against the (cheaper) abstraction
  verification can be modularized, cached and thus scaled.
- **Unbounded Verification:** Contracts can be reasoned over inductively and
  thus verify recursive functions.

Enabling function contracts is a non-invasive change. While it contains a new
API, that API is strictly additive for users. All prior operations are unchanged.

A `-Zcontracts` guard will be added wich is necessary to access any of the
contract functionality (attributes) described below. Any use of such attributes

> without `-Zcontracts` is a compile time error.

Author: Moved to header.

   inserted as `kani::assert` checks *after* the call to `my_div`. 

   The expanded version of our harness and function is equivalent to the following:

   ```rs
   #[kani::proof]
   fn my_div_harness() {
     let dividend = kani::any();
     let divisor = kani::any();
     kani::assume(divisor != 0);
     let result = my_div(dividend, divisor);
     kani::assert(result <= dividend);
   }
   ```

   This expanded harness is then verified like any other harness but also gives
   the green light for the next step: verified stubbing.

3. In the last phase the **verified** contract is ready for us to use to
   **stub** in other harnesses.

   Unlike in regular stubbing Kani there to be at least one associated
   `proof_for_contract` harness for each function to stub *and* it requires all
   such harnesses to pass verification before attempting verification of any
   harnesses that use it as a stub.

   A possible harness that uses our `my_div` contract could be the following:

   ```rs
   #[kani::proof]
   #[kani::stub_verified(my_div)]
   fn use_div() {
     let v = vec![...];
     let some_idx = my_div(v.len() - 1, 3);
     v[some_idx];
   }

>    ```

Author: Hmm, I see what you're saying. I agree that on workflow grounds it makes more sense adding another function in addition to the harness also means making it more convoluted. Mostly though I struggle in coming up with a reasonable example of a function that would use `my_div` here. I'm tabling changing this until I can find an example and I'm open to any suggestions.

   To use the contract as a stub Kani must first ensure the calling context is
   safe. It a `kani::assert` for the preconditions (`requires`) of `my_div`
   before the call. Then it replaces the result of `my_div` with a
   non-deterministic value (see [havocing](#memory-predicates-and-havocing) for
   the equivalent for mutable memory), constrained by a `kani::assume` of
   `'my_div`'s postconditions (`ensures`).

   And the expanded version is equivalent to this:
  
   ```rs
   #[kani::proof]
   #[kani::stub_verified(my_div)]
   fn use_div() {
     let v = vec![...];
     let dividend = v.len() - 1;
     let divisor = 3;
     kani::assert(divisor != 0);
     let result = kani::any();
     kani::assume(result <= dividend);
     let some_idx = result;
     v[some_idx];
   }

>    ```

Author: Hmmm, I struggle to think of a better example right now so I'm going to close this, we can always make it better later.

```rs
impl<T> Vec<T> {
  fn pop(&mut self) -> Option<T> {
    ...
  }
}
```

If we want to describe in which case the result is `Some`, we need to know
whether `self` is empty *before* `pop` is called. To do this Kani provides the
`old(EXPR)` pseudo function, which evaluates `EXPR` before the call (e.g. to
`pop`) and makes the result available to `ensures`. it would be used like so

```rs
impl<T> Vec<T> {
  #[kani::ensures(old(self.is_empty()) || result.is_some())]
  fn pop(&mut self) -> Option<T> {
    ...
  }
}
```

`old` allows evaluating any (side-effect free[^side-effects]) Rust expression
but Rust enforces that it does not borrow so as to observe the mutations from
e.g. `pop`, as that would defeat the purpose. Instead Kani encourages us to make
copies (using e.g. `clone()`) if necessary.

Note also that `old` is syntax, not a function and implemented as an extraction
and lifting during code generation. It can reference e.g. `pop`'s arguments but
not local variables. Compare the following

**Invalid ❌:** `#[kani::ensures({ let x = self.is_empty(); old(x) } || result.is_some())]`
**Valid ✅:** `#[kani::ensures(old({ let x = self.is_empty(); x }) || result.is_some())]`

> And it will only be recognized as `old(...)`, not as `let old1 = old; old1(...)` etc.

Reviewer: Shouldn't this be `kani::old()`?

> ### Memory Predicates and Havocing

Reviewer: I would actually call this something like: "Side Effects Specification"

The last new feature added are predicates to refine a function's access to heap

> memory. A memory footprint is used by the verifier to perform "havocing" during

Author: Has been rephrased

contract stubbing. Recall that stubbing replaces the result value with a
non-deterministic `kani::any()`, havocing is the equivalent memory regions
touched by the function. Any memory regions in the footprint are "havoced" by
the verifier, that is replaced by a non-deterministic value (subject to type

> constraints).

Author: I rephrased this

By default Kani infers a memory footprint as all memory reachable from a `&mut`
input or any `static` global referenced, directly or transitively, by the

> function. While the inferred footprint is sound and enough for successful

Author: I've worked around this in the text by not talking about inference but Kani's assumption given what the function would be permitted to do by the type system. I also mention that we perform checking
